pandas/tests/indexes/test_interval.py

```diff
@@ -458,7 +766,132 @@ def test_get_indexer_subintervals(self):
         expected = np.array([0, 0, 0], dtype='intp')
         tm.assert_numpy_array_equal(actual, expected)
 
+    ### NEW GET INDEXER BEHAVIOR
+
+    @pytest.mark.xfail(reason="new indexing tests for issue 16316")
+    def get_indexer_for_interval_updated_behavior(self):
+
+        index = IntervalIndex.from_tuples([(0, 2.5), (1, 3), (2, 4)], closed='right')
+
+        # single queries
+        result = index.get_indexer([Interval(1, 3, closed='right')])
+        expected = np.array([1], dtype='intp')
+        tm.assert_numpy_array_equal(result, expected)
+
+        result = index.get_indexer([Interval(1, 3, closed='left')])
+        expected = np.array([-1], dtype='intp')
+        tm.assert_numpy_array_equal(result, expected)
+
+        result = index.get_indexer([Interval(1, 3, closed='both')])
+        expected = np.array([-1], dtype='intp')
+        tm.assert_numpy_array_equal(result, expected)
+
+        result = index.get_indexer([Interval(1, 3, closed='neither')])
+        expected = np.array([-1], dtype='intp')
+        tm.assert_numpy_array_equal(result, expected)
+
+        result = index.get_indexer([Interval(1, 4, closed='right')])
+        expected = np.array([-1], dtype='intp')
+        tm.assert_numpy_array_equal(result, expected)
+
+        result = index.get_indexer([Interval(0, 4, closed='right')])
+        expected = np.array([-1], dtype='intp')
+        tm.assert_numpy_array_equal(result, expected)
+
+        result = index.get_indexer([Interval(1, 2, closed='right')])
+        expected = np.array([-1], dtype='intp')
+        tm.assert_numpy_array_equal(result, expected)
+
+        # multiple queries
+        result = index.get_indexer([Interval(2, 4, closed='right'),
+                                    Interval(1, 3, closed='right') ])
+        expected = np.array([2, 1], dtype='intp')
+        tm.assert_numpy_array_equal(result, expected)
+
+        result = index.get_indexer([Interval(1, 3, closed='right'),
+                                    Interval(0, 2, closed='right') ])
+        expected = np.array([1, -1], dtype='intp')
+        tm.assert_numpy_array_equal(result, expected)
+
+        result = index.get_indexer([Interval(1, 3, closed='right'),
+                                    Interval(1, 3, closed='left') ])
+        expected = np.array([1, -1], dtype='intp')
+        tm.assert_numpy_array_equal(result, expected)
+
+        result = index.get_indexer(index)
+        expected = np.array([0, 1, 2], dtype='intp')
+        tm.assert_numpy_array_equal(result, expected)
```

Comment:
I would add a test case with duplicates in the index as well (which should raise an error based on the discussion with @shoyer), and then actually for `get_indexer_non_unique` as well

```
In [40]: pd.Index([1,1,3,2]).get_indexer([1])
...
InvalidIndexError: Reindexing only valid with uniquely valued Index objects

In [41]: pd.Index([1,1,3,2]).get_indexer_non_unique([1])
Out[41]: (Int64Index([0, 1], dtype='int64'), array([], dtype=int64))
```